agent/engine/docker_events_buffer.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package engine
+
+import (
+	docker "github.com/fsouza/go-dockerclient"
+	"sync"
+)
+
+const (
+	containerTypeEvent = "container"
+)
+
+var ignoredDockerEvents = []string{"rename", "kill", "destroy", "pause", "exec_create", "exec_start", "top", "attach", "export", "pull", "push", "tag", "untag", "import", "delete"}
+
+// InfiniteBuffer defines an unlimited buffer, where it reads from
+// input channel and write to output channel.
+type InfiniteBuffer struct {
+	events   []*docker.APIEvents
+	empty    bool
+	waitDone chan struct{}
+	count    int
+	lock     sync.RWMutex
+}
+
+// NewInfiniteBuffer returns an InfiniteBuffer object
+func NewInfiniteBuffer() *InfiniteBuffer {
+	return &InfiniteBuffer{
+		waitDone: make(chan struct{}),
+	}
+}
+
+// Serve starts reading from the input channel and writes to the buffer
+func (buffer *InfiniteBuffer) Serve(events chan *docker.APIEvents) {
```

Comment:
The very common `Serve()` method from `http` is a blocking call and it might be confusing for us to have a method named `Serve` that's not blocking. Can you rename this to `StartListener()` or something?